pkg/controller/v1alpha2/core/scopes/healthscope/healthscope_controller_test.go

```diff
@@ -0,0 +1,343 @@
+	fieldpath.Pave(hGeneralWL.Object).SetValue("metadata.name", "healthyGeneral") // healthy
+
+	uhGeneralRef := v1alpha1.TypedReference{
+		APIVersion: "unknown",
+		Kind:       "unknown",
+		Name:       "unhealthyGeneral",
+	}
+	uhDeploy := hDeploy
+	uhDeploy.Status.ReadyReplicas = 0 // unhealthy
+	uhGeneralWL := &unstructured.Unstructured{Object: make(map[string]interface{})}
+	fieldpath.Pave(uhGeneralWL.Object).SetValue("status.readyReplicas", 0)           // healthy
+	fieldpath.Pave(uhGeneralWL.Object).SetValue("metadata.name", "unhealthyGeneral") // healthy
+	unsupporttedWL := &unstructured.Unstructured{Object: make(map[string]interface{})}
+	fieldpath.Pave(unsupporttedWL.Object).SetValue("status.unknown", 1) // healthy
+
+	BeforeEach(func() {
+		logf.Log.Info("Set up resources before an unit test")
+		hs.Spec.WorkloadReferences = []v1alpha1.TypedReference{}
+	})
+
+	AfterEach(func() {
+		logf.Log.Info("Clean up resources after an unit test")
+	})
+
+	// use ContainerizedWorkload and Deployment checker
+	It("Test healthy scope", func() {
+		tests := []struct {
+			caseName       string
+			hsWorkloadRefs []v1alpha1.TypedReference
+			mockGetFn      test.MockGetFn
+		}{
+			{
+				caseName:       "2 supportted workloads(cw,deploy)",
+				hsWorkloadRefs: []v1alpha1.TypedReference{cwRef, deployRef},
+				mockGetFn: func(ctx context.Context, key types.NamespacedName, obj runtime.Object) error {
+					if o, ok := obj.(*corev1alpha2.ContainerizedWorkload); ok {
+						*o = cw
+					}
+					if o, ok := obj.(*appsv1.Deployment); ok {
+						*o = hDeploy
+					}
+					return nil
+				},
+			},
+			{
+				caseName:       "1 supportted workload and 1 general workload",
+				hsWorkloadRefs: []v1alpha1.TypedReference{cwRef, hGeneralRef},
+				mockGetFn: func(ctx context.Context, key types.NamespacedName, obj runtime.Object) error {
+					switch o := obj.(type) {
+					case *corev1alpha2.ContainerizedWorkload:
+						*o = cw
+					case *appsv1.Deployment:
+						*o = hDeploy
+					case *unstructured.Unstructured:
+						*o = *hGeneralWL
+					}
+					return nil
+				},
+			},
+			{
+				caseName:       "2 general workloads",
+				hsWorkloadRefs: []v1alpha1.TypedReference{hGeneralRef, hGeneralRef},
+				mockGetFn: func(ctx context.Context, key types.NamespacedName, obj runtime.Object) error {
+					o, _ := obj.(*unstructured.Unstructured)
+					*o = *hGeneralWL
+					return nil
+				},
+			},
+		}
+		for _, tc := range tests {
+			By("Running: " + tc.caseName)
+			mockClient := &test.MockClient{
+				MockGet: tc.mockGetFn,
+			}
+			reconciler.client = mockClient
+			hs.Spec.WorkloadReferences = tc.hsWorkloadRefs
+			result := reconciler.GetScopeHealthStatus(ctx, &hs)
+			Expect(result).ShouldNot(BeNil())
+			Expect(result.IsHealthy).Should(Equal(true))
+		}
+		//TODO(roywang) golang question:
+		// as shown below
+		// why tc.hsWorkloadRefs passed into It func
+		// shared the same address?
+		// for _, tc := range tests {
+		//
```

Comment:
`tc` is a local variable in loop, and the address behind it is an iterator. So when the loop iterated, the value of iterator which is `tc` will change while the address keep the same. Generally, add such line in loop will fix:

```
tc := tc
```

Reply:
Yes, I used `tc:=tc` originally but it didn't work because I put `tc:=tc` inside `It("",func(){tc:=tc ...})`. It should put it before `It` otherwise the scope of `It` will always get the same address of `tc`. Fixed.  Thx!